### PR TITLE
initial implementation of ASYNC123

### PR DESCRIPTION
**Pull request by jakkdl**
messy and imperfect, but should catch most usage and hopefully Ruff can implement this properly :crossed_fingers: 

TODO:
* changelog, bump version
* write documentation
* figure out a name. `reraise_child_exception` is clean but "child exception" is not established nomenclature. `reraise_exception_group_sub_exception`? `bad_exception_group_flattening`?

### CODE REVIEW

File: docs/rules.rst [hunk 1]
```diff
@@ -89,6 +89,9 @@ _`ASYNC121`: control-flow-in-taskgroup
 _`ASYNC122`: delayed-entry-of-relative-cancelscope
     :func:`trio.move_on_after`, :func:`trio.fail_after`, :func:`anyio.move_on_after` and :func:`anyio.fail_after` behaves unintuitively if initialization and entry are separated, with the timeout starting on initialization. Trio>=0.27 changes this behaviour, so if you don't support older versions you should disable this check. See `Trio issue #2512 <https://github.com/python-trio/trio/issues/2512>`_.
 
+_`ASYNC123`: bad-exception-group-flattening
+    Raising an exception that was inside an `ExceptionGroup` inside the ``except`` handler of that group will set the group as the `__context__` for the exception, which overrides the previous `__context__` that exception had, truncating the context tree. The same is true for ``__cause__`` (unless explicitly setting ``from``) and `__traceback__`. The easiest fix is to `copy.copy` the exception before raising it, but beware that `trio.Cancelled` did not support copying before `trio==0.27.1`.
```
Comment (Zac-HD) [inline docs/rules.rst:93]:
```suggestion
    Raising one of the contained exceptions will mutate it, replacing the original ``.__context__`` with the group, and erasing the ``.__traceback__``.
    Dropping this information makes diagnosing errors much more difficult.
    We recommend ``raise SomeNewError(...) from group`` if possible; or consider using `copy.copy` to shallow-copy the exception before re-raising (for copyable types), or re-raising the error from outside the `except` block.
```

**Review by Zac-HD — APPROVED** [review]
Implementation and tests look great, thanks @jakkdl!  

I've suggested some changes to the docs which I think will give a bit more context on what happens and why it's bad, as well as recommending "don't do this at all" as our top suggestion.  Merge when you're happy with it, with or without edits 🙂

File: docs/rules.rst [hunk 2]
```diff
@@ -89,6 +89,11 @@ _`ASYNC121`: control-flow-in-taskgroup
 _`ASYNC122`: delayed-entry-of-relative-cancelscope
     :func:`trio.move_on_after`, :func:`trio.fail_after`, :func:`anyio.move_on_after` and :func:`anyio.fail_after` behaves unintuitively if initialization and entry are separated, with the timeout starting on initialization. Trio>=0.27 changes this behaviour, so if you don't support older versions you should disable this check. See `Trio issue #2512 <https://github.com/python-trio/trio/issues/2512>`_.
 
+_`ASYNC123`: bad-exception-group-flattening
+    Raising one of the contained exceptions will mutate it, replacing the original ``.__context__`` with the group, and erasing the ``.__traceback__``.
```
Comment (jakkdl) [inline docs/rules.rst:93]:
```suggestion
    Raising one of the exceptions contained in an exception group will mutate it, replacing the original ``.__context__`` with the group, and erasing the ``.__traceback__``.
```

File: docs/rules.rst (hunk 1 above)
Comment (jakkdl) [reply docs/rules.rst]: That's much better, thanks! Just one small nit on it

**Review by Zac-HD — APPROVED** [review]

**Pull request merged**